File: 03/src/main/java/ru/otus/spring/hw3/game/questions/csv/CSVQuestionsDao.java
package ru.otus.spring.hw3.game.questions.csv;

import com.opencsv.CSVReader;
import org.springframework.core.io.Resource;
import org.springframework.stereotype.Service;
import ru.otus.spring.hw3.game.questions.Question;
import ru.otus.spring.hw3.game.questions.QuestionsDao;

import java.io.FileReader;
import java.io.IOException;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@Service
public class CSVQuestionsDao implements QuestionsDao {
    private List<Question> questions;

    public CSVQuestionsDao(Resource gameFile) throws IOException {
        try (
                FileReader fileReader = new FileReader(gameFile.getFile());
                CSVReader csvReader = new CSVReader(fileReader)
        ) {
            questions = readQuestions(csvReader);
        }
    }

    private List<Question> readQuestions(Iterable<String[]> questions) {
        return StreamSupport.stream(questions.spliterator(), false)
                .map(CSVQuestionsDao::convertLineToQuestion)
                .collect(Collectors.toList());
    }

    private static Question convertLineToQuestion(String[] line) {
        String question = line[0];
        String answer = line[1];
        return new Question(question, answer);
    }

    @Override
    public Question getQuestion(int questionIndex) {
        return questions.get(questionIndex);
    }

    @Override
    public int getQuestionCount() {
        return questions.size();
    }
}

File: 04/src/main/kotlin/ru/otus/spring/hw4/questions/csv/CSVQuestionsDAO.kt
package ru.otus.spring.hw4.questions.csv

import com.opencsv.CSVReader
import org.springframework.core.io.Resource
import org.springframework.stereotype.Service
import ru.otus.spring.hw4.Question
import ru.otus.spring.hw4.questions.QuestionsDAO
import java.io.FileReader

@Service
class CSVQuestionsDAO(
        gameFile: Resource
) : QuestionsDAO {
    private lateinit var questions: List<Question>

    init {
        val lines = FileReader(gameFile.file).use { fileReader ->
            CSVReader(fileReader).use {
                it.readAll()
            }
        }
        questions = readQuestions(lines)
    }

    private fun readQuestions(questionLines: List<Array<String>>): List<Question> {
        return questionLines.mapNotNull(this::convertLineToQuestion)
    }

    private fun convertLineToQuestion(line: Array<String>): Question? {
        val question = line[0]
        val answer = line[1]
        return Question(question, answer)
    }

    override fun getQuestion(questionIndex: Int): Question {
        return questions[questionIndex]
    }

    override fun getQuestionCount(): Int {
        return questions.size
    }

}

